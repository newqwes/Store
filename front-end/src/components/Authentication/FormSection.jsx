import React from 'react';
import PropTypes from 'prop-types';
import { NavLink, Redirect } from 'react-router-dom';

import contentType from './propTypes';

import ROUTER_PATH from '../../constants/routerPath';
import THEME_VARIANT from '../../constants/themeVariant';
import { ALIGN_ITEMS, JUSTIFY_CONTENT } from '../../constants/position';

import Flex from '../Flex';
import Label from '../Label';
import Button from '../Button';
import FieldList from '../FieldList';
import InputField from '../InputField';

import { FormWrapper, FormContent, InputFieldContent } from './styled';

class FormSection extends React.Component {
  state = {
    isErrorAnimation: false,
  };

  onAnimationEnd = () => this.setState({ isErrorAnimation: false });

  handleClick = () => this.setState({ isErrorAnimation: true });

  static propTypes = {
    themeVariant: PropTypes.string,
    submit: PropTypes.func.isRequired,
    userAuthorized: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired,
    content: PropTypes.shape(contentType),
  };

  static defaultProps = {
    themeVariant: THEME_VARIANT.default,
  };

  render() {
    const {
      themeVariant,
      submit,
      handleSubmit,
      userAuthorized,
      content: { title, buttonText, link, fields },
    } = this.props;
    const { isErrorAnimation } = this.state;

    if (userAuthorized) return <Redirect to={ROUTER_PATH.products.pizza} />;

    return (
      <FormWrapper themeVariant={themeVariant}>
        <Flex alignItems={ALIGN_ITEMS.center}>
          <FormContent
            themeVariant={themeVariant}
            isErrorAnimation={isErrorAnimation}
            onAnimationEnd={this.onAnimationEnd}>
            <Flex justifyContent={JUSTIFY_CONTENT.center}>
              <Label text={title} />
            </Flex>
            <form onSubmit={handleSubmit(submit)}>
              <FieldList fieldStyle={InputFieldContent} fields={fields} component={InputField} />
              <Button text={buttonText} onClick={this.handleClick} />
              <Flex justifyContent={JUSTIFY_CONTENT.center}>
                <NavLink to={link.route}>{link.text}</NavLink>
              </Flex>
            </form>
          </FormContent>
        </Flex>
      </FormWrapper>
    );
  }
}

export default FormSection;
