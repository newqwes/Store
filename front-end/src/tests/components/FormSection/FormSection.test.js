import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { reduxForm, reducer as formReducer } from 'redux-form';
import { createStore, combineReducers } from 'redux';
import { Provider } from 'react-redux';

import { mountWithTheme } from '../../helpers';
import { mockAuthenticationForm } from '../../helpers/mocks';

import {
  FormWrapper,
  FormContent,
  InputFieldContent,
} from '../../../components/Authentication/styled';

import FormSection from '../../../components/Authentication';

const rootReducer = combineReducers({
  form: formReducer,
});

const store = createStore(rootReducer);

const FormSectionWithReduxForm = reduxForm({
  form: 'login',
})(FormSection);

const setUp = props =>
  mountWithTheme(
    <Router>
      <Provider store={store}>
        <FormSectionWithReduxForm
          content={mockAuthenticationForm}
          submit={() => {}}
          handleSubmit={() => {}}
          userAuthorized={false}
          {...props}
        />
      </Provider>
    </Router>,
  );

describe('snapshot FormSection', () => {
  test('snapshot', () => {
    const wrapper = setUp();

    expect(wrapper).toMatchSnapshot();
  });
});

describe('FormSection component', () => {
  const mockSubmit = jest.fn();
  const mockHandleSubmit = jest.fn();

  const wrapper = setUp({ submit: mockSubmit, handleSubmit: mockHandleSubmit });

  it(`should render FormSection with following children:
  - 1 FormWrapper,
  - 1 FormContent,
  - 2 InputFieldContent,
  - 3 Flex,
  - 1 Label,
  - 1 NavLink,
  - 1 form,
  - 2 Button`, () => {
    const formWrapper = wrapper.find(FormWrapper);
    expect(formWrapper.length).toBe(1);

    const formContent = wrapper.find(FormContent);
    expect(formContent.length).toBe(1);

    const inputFieldContent = wrapper.find(InputFieldContent);
    expect(inputFieldContent.length).toBe(2);

    const flexs = wrapper.find('Flex');
    expect(flexs.length).toBe(3);

    const labels = wrapper.find('Label');
    expect(labels.length).toBe(1);

    const buttons = wrapper.find('Button');
    expect(buttons.length).toBe(1);

    const form = wrapper.find('form');
    expect(form.length).toBe(1);

    const navLink = wrapper.find('NavLink');
    expect(navLink.length).toBe(1);
  });

  it('should be animation when click submit', () => {
    const formSection = wrapper.find('FormSection');
    const buttons = wrapper.find('Button');

    buttons.simulate('click');
    expect(formSection.state()).toEqual({ isErrorAnimation: true });
  });

  it('should be submit when click button', () => {
    const buttons = wrapper.find('Button');
    buttons.simulate('click');

    expect(mockHandleSubmit).toHaveBeenCalledTimes(1);
  });

  it('should end animation when call onAnimationEnd', () => {
    const formSection = wrapper.find('FormSection');
    const buttons = wrapper.find('Button');

    buttons.simulate('click');
    formSection.instance().onAnimationEnd();
    expect(formSection.state()).toEqual({ isErrorAnimation: false });
  });

  it('should redirect when user authorized', () => {
    const formWrapper = setUp({ userAuthorized: true }).find(FormWrapper);
    expect(formWrapper.length).toBe(0);
  });
});
