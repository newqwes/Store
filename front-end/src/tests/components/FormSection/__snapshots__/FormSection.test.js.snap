// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot FormSection snapshot 1`] = `
<section
  className="sc-bdfBwQ jsNWCE"
>
  <div
    className="sc-hKgILt kbpzhu"
    direction="row"
  >
    <div
      className="sc-gsTCUz jwhmQt"
      onAnimationEnd={[Function]}
    >
      <div
        className="sc-hKgILt dElrSb"
        direction="row"
      >
        <span
          className="sc-eCssSg gpgRrO"
          fontSize={20}
          fontWeight={700}
        >
          Авторизация
        </span>
      </div>
      <form>
        Array [
          <div
            className="sc-dlfnbm dJMQwA"
            disabled={false}
          >
            <label>
              Почта
            </label>
            <input
              disabled={false}
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              placeholder="example@gmail.com"
              type="text"
              value=""
            />
          </div>,
          <div
            className="sc-dlfnbm dJMQwA"
            disabled={false}
          >
            <label>
              Пароль
            </label>
            <input
              disabled={false}
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              placeholder="Не менее 5 символов"
              type="password"
              value=""
            />
          </div>,
        ]
        <button
          className="sc-jSgupP fNTAJX"
          onClick={[Function]}
        >
          Войти
        </button>
        <div
          className="sc-hKgILt dElrSb"
          direction="row"
        >
          <a
            aria-current={null}
            href="/registration"
            onClick={[Function]}
          >
            Регистрация
          </a>
        </div>
      </form>
    </div>
  </div>
</section>
`;
